# Logestic Regression Model - COEF_features

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sqlalchemy import create_engine
from config import db_password
import matplotlib.pyplot as plt
import seaborn as sns
import math

In [2]:
db_string = f"postgresql://postgres:{db_password}@127.0.0.1:5433/nba_champs"
engine = create_engine(db_string)

In [3]:
stats_df = pd.read_sql("select * from \"stats\"", db_string);
stats_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [4]:
test_df= pd.read_sql("select * from \"cy_stats\"", db_string)
test_df.head()

index  Position   Season   Tm  G  W  L  Win_pct    MOV   SOS  ...  eFG_pct  \
0      0         0  2021-22  MIA  6  5  1    0.833  17.00  1.07  ...    0.525   
1      1         0  2021-22  WAS  7  5  2    0.714   2.57  1.11  ...    0.505   
2      2         0  2021-22  UTA  6  5  1    0.833  12.83 -3.08  ...    0.526   
3      3         0  2021-22  TOR  8  5  3    0.625   4.63  1.25  ...    0.487   
4      4         0  2021-22  SAS  7  2  5    0.286  -0.43 -1.50  ...    0.526   

   TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0   0.565     13.6     27.5   0.197        0.464       0.501         13.4   
1   0.549     11.5     20.4   0.213        0.492       0.535         11.0   
2   0.574     14.2     24.6   0.236        0.450       0.485         10.8   
3   0.518     12.1     29.8   0.158        0.521       0.554         16.8   
4   0.542     12.2     22.2   0.113        0.549       0.568         15.2   

   Opp_ORB_pct  Opp_FT_FGA  
0         15.9       0.181  
1         20.5       0.197  
2         23.1       0.164  
3         20.6       0.178  
4         21.4       0.130  

[5 rows x 24 columns]

In [5]:
# Pull out Season, TM and Team_Name into DF
name_df = stats_df[['Season', 'Tm',]]
name_df.head()

Season   Tm
0  2020-21  WAS
1  2020-21  UTA
2  2020-21  TOR
3  2020-21  SAS
4  2020-21  SAC

In [6]:
#list(stats_df.columns)

In [7]:
all_features_df = stats_df
all_features_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [8]:
stats_df.head()

index  Position   Season   Tm   G   W   L  Win_pct   MOV   SOS  ...  \
0      0         0  2020-21  WAS  72  34  38    0.472 -1.83 -0.01  ...   
1      1         0  2020-21  UTA  72  52  20    0.722  9.25 -0.29  ...   
2      2         0  2020-21  TOR  72  27  45    0.375 -0.47 -0.07  ...   
3      3         0  2020-21  SAS  72  33  39    0.458 -1.74  0.15  ...   
4      4         0  2020-21  SAC  72  31  41    0.431 -3.68  0.23  ...   

   eFG_pct  TS_pct  TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  \
0    0.531   0.569     12.3     21.3   0.221        0.539       0.576   
1    0.563   0.597     12.7     24.5   0.195        0.507       0.537   
2    0.529   0.567     11.9     20.8   0.196        0.543       0.584   
3    0.517   0.554     10.2     20.0   0.192        0.541       0.570   
4    0.549   0.578     12.0     21.3   0.185        0.557       0.591   

   Opp_TOV_pct  Opp_ORB_pct  Opp_FT_FGA  
0         12.5         22.4       0.217  
1         10.3         20.7       0.159  
2         14.4         23.7       0.234  
3         11.8         22.7       0.174  
4         12.1         25.0       0.199  

[5 rows x 24 columns]

In [9]:
# Drop un needed columns
all_features_df.drop(['index',
 'Season',
 'Tm',
 'G',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 #'Pace',
 #'ORtg',
 #'DRtg',
 #'eFG_pct',
 #'TS_pct',
 #'TOV_pct',
 #'ORB_pct',
 #'FT_FGA',
 #'Opp_eFG_pct',
 #'Opp_TS_pct',
 #'Opp_TOV_pct',
 #'Opp_ORB_pct',
 #'Opp_FT_FGA'
              ], axis=1, inplace=True)

In [10]:
all_features_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg   DRtg  eFG_pct  TS_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2  113.0    0.531   0.569   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6  108.3    0.563   0.597   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0  112.5    0.529   0.567   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0  112.8    0.517   0.554   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6  117.2    0.549   0.578   

   TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0     12.3     21.3   0.221        0.539       0.576         12.5   
1     12.7     24.5   0.195        0.507       0.537         10.3   
2     11.9     20.8   0.196        0.543       0.584         14.4   
3     10.2     20.0   0.192        0.541       0.570         11.8   
4     12.0     21.3   0.185        0.557       0.591         12.1   

   Opp_ORB_pct  Opp_FT_FGA  
0         22.4       0.217  
1         20.7       0.159  
2         23.7       0.234  
3         22.7       0.174  
4         25.0       0.199

In [11]:
# Remove Position target from features data
y = all_features_df.Position
X = all_features_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)


In [12]:
feature_names = X.columns

In [13]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [14]:
# Create a logistic regression model
classifier_AF = LogisticRegression(max_iter=1000)
classifier_AF

LogisticRegression(max_iter=1000)

In [15]:
# Fit (train) our model by using the training data
classifier_AF.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [16]:
# Validate the model by using the test data
print(f"Logistic Regression Training Data Score: {classifier_AF.score(X_train_scaled, y_train)}")
print(f"Logistic Regression Testing Data Score: {classifier_AF.score(X_test_scaled, y_test)}")

Logistic Regression Training Data Score: 0.9642346208869814
Logistic Regression Testing Data Score: 0.9786324786324786


All features score = 0.97863

Find accuracy using top 10 coefficient features

In [17]:
# https://towardsdatascience.com/a-look-into-feature-importance-in-logistic-regression-models-a4aa970f9b0f
import numpy as np
feature_bol = np.round(classifier_AF.coef_,decimals=2) 
feature_bol

array([[ 1.06,  0.19,  0.35,  0.22,  0.02,  0.11, -0.41,  1.05,  0.  ,
        -0.13,  0.76, -0.51, -0.6 , -0.17,  0.05,  0.01, -0.24]])

In [18]:
feature_names

Index(['Win_pct', 'MOV', 'SOS', 'SRS', 'Pace', 'ORtg', 'DRtg', 'eFG_pct',
       'TS_pct', 'TOV_pct', 'ORB_pct', 'FT_FGA', 'Opp_eFG_pct', 'Opp_TS_pct',
       'Opp_TOV_pct', 'Opp_ORB_pct', 'Opp_FT_FGA'],
      dtype='object')

In [19]:
coef_features_df = all_features_df
coef_features_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg   DRtg  eFG_pct  TS_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2  113.0    0.531   0.569   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6  108.3    0.563   0.597   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0  112.5    0.529   0.567   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0  112.8    0.517   0.554   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6  117.2    0.549   0.578   

   TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0     12.3     21.3   0.221        0.539       0.576         12.5   
1     12.7     24.5   0.195        0.507       0.537         10.3   
2     11.9     20.8   0.196        0.543       0.584         14.4   
3     10.2     20.0   0.192        0.541       0.570         11.8   
4     12.0     21.3   0.185        0.557       0.591         12.1   

   Opp_ORB_pct  Opp_FT_FGA  
0         22.4       0.217  
1         20.7       0.159  
2         23.7       0.234  
3         22.7       0.174  
4         25.0       0.199

In [20]:
stats_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg   DRtg  eFG_pct  TS_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2  113.0    0.531   0.569   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6  108.3    0.563   0.597   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0  112.5    0.529   0.567   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0  112.8    0.517   0.554   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6  117.2    0.549   0.578   

   TOV_pct  ORB_pct  FT_FGA  Opp_eFG_pct  Opp_TS_pct  Opp_TOV_pct  \
0     12.3     21.3   0.221        0.539       0.576         12.5   
1     12.7     24.5   0.195        0.507       0.537         10.3   
2     11.9     20.8   0.196        0.543       0.584         14.4   
3     10.2     20.0   0.192        0.541       0.570         11.8   
4     12.0     21.3   0.185        0.557       0.591         12.1   

   Opp_ORB_pct  Opp_FT_FGA  
0         22.4       0.217  
1         20.7       0.159  
2         23.7       0.234  
3         22.7       0.174  
4         25.0       0.199

In [21]:
# Drop un needed columns
coef_features_df.drop([#'index',
 #'Season',
 #'Tm',
 #'G',
 #'W',
 #'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 #'Pace',
 #'ORtg',
 'DRtg',
 #'eFG_pct',
 'TS_pct',
 'TOV_pct',
 #'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TS_pct',
 #'Opp_TOV_pct',
 #'Opp_ORB_pct',
 'Opp_FT_FGA'
              ], axis=1, inplace=True)

In [22]:
stats_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg  eFG_pct  ORB_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2    0.531     21.3   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6    0.563     24.5   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0    0.529     20.8   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0    0.517     20.0   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6    0.549     21.3   

   Opp_TOV_pct  Opp_ORB_pct  
0         12.5         22.4  
1         10.3         20.7  
2         14.4         23.7  
3         11.8         22.7  
4         12.1         25.0

In [23]:
all_features_df.head()

Position  Win_pct   MOV   SOS   SRS   Pace   ORtg  eFG_pct  ORB_pct  \
0         0    0.472 -1.83 -0.01 -1.85  104.1  111.2    0.531     21.3   
1         0    0.722  9.25 -0.29  8.97   98.5  117.6    0.563     24.5   
2         0    0.375 -0.47 -0.07 -0.54   99.2  112.0    0.529     20.8   
3         0    0.458 -1.74  0.15 -1.58   98.9  111.0    0.517     20.0   
4         0    0.431 -3.68  0.23 -3.45  100.0  113.6    0.549     21.3   

   Opp_TOV_pct  Opp_ORB_pct  
0         12.5         22.4  
1         10.3         20.7  
2         14.4         23.7  
3         11.8         22.7  
4         12.1         25.0

In [24]:
# Remove Position target from features data
y = coef_features_df.Position
X = coef_features_df.drop(columns=["Position"])

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

In [25]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Scale the data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [26]:
# Create a logistic regression model
classifier_CF = LogisticRegression(max_iter=1000)
classifier_CF

LogisticRegression(max_iter=1000)

In [27]:
# Fit (train) our model by using the training data
classifier_CF.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [28]:
# Validate the model by using the test data
print(f"Logistic Regression with coef features only Training Data Score: {classifier_CF.score(X_train_scaled, y_train)}")
print(f"Logistic Regression with coef features only Testing Data Score: {classifier_CF.score(X_test_scaled, y_test)}")

Logistic Regression with coef features only Training Data Score: 0.9670958512160229
Logistic Regression with coef features only Testing Data Score: 0.9743589743589743


Coefficient features score = 0.97435

In [29]:
# Pull out Season, TM and Team_Name into DF
test_name_df = test_df[['Season', 'Tm']]
test_name_df.head()

Season   Tm
0  2021-22  MIA
1  2021-22  WAS
2  2021-22  UTA
3  2021-22  TOR
4  2021-22  SAS

In [30]:
test_df_pos= test_df['Position']
test_df.drop(['index',
 'Position',
 'Season',
 'Tm',
 'G',
 'W',
 'L',
 #'Win_pct',
 #'MOV',
 #'SOS',
 #'SRS',
 #'Pace',
 #'ORtg',
 'DRtg',
 #'eFG_pct',
 'TS_pct',
 'TOV_pct',
 #'ORB_pct',
 'FT_FGA',
 'Opp_eFG_pct',
 'Opp_TS_pct',
 #'Opp_TOV_pct',
 #'Opp_ORB_pct',
 'Opp_FT_FGA'
             ], axis=1, inplace=True)
test_df

Win_pct    MOV   SOS    SRS   Pace   ORtg  eFG_pct  ORB_pct  Opp_TOV_pct  \
0     0.833  17.00  1.07  18.07   99.9  112.2    0.525     27.5         13.4   
1     0.714   2.57  1.11   3.68   99.0  108.8    0.505     20.4         11.0   
2     0.833  12.83 -3.08   9.76   99.3  111.9    0.526     24.6         10.8   
3     0.625   4.63  1.25   5.88   97.6  107.4    0.487     29.8         16.8   
4     0.286  -0.43 -1.50  -1.93  101.7  106.6    0.526     22.2         15.2   
5     0.500  -2.83 -1.06  -3.89  101.1  107.8    0.508     21.7         10.1   
6     0.429   1.86 -0.95   0.91  100.6  111.3    0.540     25.9         12.2   
7     0.400  -5.00 -1.64  -6.64  100.8  104.4    0.511     20.4         13.4   
8     0.714   8.57 -2.13   6.44   96.3  116.4    0.571     19.0         13.0   
9     0.250  -7.63  3.26  -4.37   98.1  105.9    0.518     24.4         10.8   
10    0.143 -13.14  0.10 -13.05   99.6   97.1    0.472     18.3         12.9   
11    0.714   4.86  0.52   5.38   98.1  114.0    0.558     20.6         11.6   
12    0.143  -6.71  1.23  -5.48   99.7  104.6    0.491     26.4         12.2   
13    0.500   0.17 -4.22  -4.06  100.3  102.9    0.486     26.4         16.6   
14    0.429  -3.57  3.87   0.30  102.0  106.1    0.509     20.1         12.2   
15    0.571   1.29 -0.98   0.31   97.8  110.2    0.507     27.9         10.6   
16    0.571  -2.86  1.75  -1.10   99.4  110.3    0.512     28.6         13.0   
17    0.571   0.57 -4.88  -4.31  103.1  107.6    0.542     20.8         14.4   
18    0.333  -0.83 -2.11  -2.94  101.0  101.2    0.486     18.1         16.3   
19    0.250  -2.00  5.05   3.05  100.6  107.8    0.538     22.3         12.7   
20    0.167  -7.67 -3.49 -11.15  103.1   98.5    0.496     21.9         14.1   
21    0.833   7.83 -6.39   1.45  101.8  107.0    0.538     20.3         13.5   
22    0.167 -11.00  3.91  -7.09   99.2   96.9    0.464     22.6         16.2   
23    0.571   2.57 -1.72   0.85   99.1  101.6    0.536     18.5         15.0   
24    0.667  -4.67 -1.66  -6.32   98.6   99.7    0.477     20.1         13.5   
25    0.500  -1.13 -0.62  -1.75   99.0  104.9    0.499     25.4         14.1   
26    0.625   2.50  1.92   4.42  100.9  114.9    0.537     26.0         14.0   
27    0.857   8.71 -0.23   8.49   99.2  110.6    0.530     17.3         15.1   
28    0.571   0.00  4.13   4.13  100.5  104.1    0.530     15.9         10.7   
29    0.286  -5.86  2.91  -2.95   99.9  106.1    0.501     20.3         13.2   

    Opp_ORB_pct  
0          15.9  
1          20.5  
2          23.1  
3          20.6  
4          21.4  
5          24.7  
6          19.1  
7          18.7  
8          26.3  
9          25.5  
10         22.8  
11         23.2  
12         21.6  
13         30.5  
14         20.1  
15         20.8  
16         21.5  
17         26.2  
18         25.8  
19         22.4  
20         21.8  
21         15.8  
22         18.7  
23         21.4  
24         22.2  
25         27.2  
26         25.9  
27         25.1  
28         21.0  
29         24.2

In [31]:
df_X_test_scaled = scaler.fit_transform(test_df)

In [32]:
df_test_df=classifier_CF.predict(df_X_test_scaled)
df_test_df

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [33]:
cy_test_df = pd.DataFrame(df_test_df, columns = ['Position'])
cy_test_df.head()

Position
0         1
1         0
2         0
3         0
4         0

In [34]:
joined_df = test_name_df.join(cy_test_df)
joined_df

Season   Tm  Position
0   2021-22  MIA         1
1   2021-22  WAS         0
2   2021-22  UTA         0
3   2021-22  TOR         0
4   2021-22  SAS         0
5   2021-22  SAC         0
6   2021-22  POR         0
7   2021-22  PHO         0
8   2021-22  PHI         0
9   2021-22  ORL         0
10  2021-22  OKC         0
11  2021-22  NYK         0
12  2021-22  NOP         0
13  2021-22  MIN         0
14  2021-22  MIL         0
15  2021-22  ATL         0
16  2021-22  MEM         0
17  2021-22  LAL         0
18  2021-22  LAC         0
19  2021-22  IND         0
20  2021-22  HOU         0
21  2021-22  GSW         0
22  2021-22  DET         0
23  2021-22  DEN         0
24  2021-22  DAL         0
25  2021-22  CLE         0
26  2021-22  CHO         0
27  2021-22  CHI         0
28  2021-22  BRK         0
29  2021-22  BOS         0

In [36]:
classifier_CF.predict_proba(df_X_test_scaled)

array([[3.22571532e-01, 6.77428468e-01],
       [9.81949595e-01, 1.80504047e-02],
       [7.31351361e-01, 2.68648639e-01],
       [9.76450522e-01, 2.35494777e-02],
       [9.99367729e-01, 6.32270654e-04],
       [9.98390841e-01, 1.60915860e-03],
       [9.93935970e-01, 6.06402974e-03],
       [9.99630896e-01, 3.69103580e-04],
       [9.48269087e-01, 5.17309130e-02],
       [9.99144910e-01, 8.55089961e-04],
       [9.99985375e-01, 1.46246643e-05],
       [9.43342333e-01, 5.66576674e-02],
       [9.99813119e-01, 1.86881169e-04],
       [9.98537770e-01, 1.46222969e-03],
       [9.98163928e-01, 1.83607210e-03],
       [9.89096980e-01, 1.09030203e-02],
       [9.93194291e-01, 6.80570931e-03],
       [9.97308613e-01, 2.69138688e-03],
       [9.99774614e-01, 2.25385786e-04],
       [9.96992506e-01, 3.00749352e-03],
       [9.99944528e-01, 5.54715139e-05],
       [9.58829975e-01, 4.11700250e-02],
       [9.99939765e-01, 6.02354296e-05],
       [9.88591260e-01, 1.14087398e-02],
       [9.984252